base/subarray.jl

```diff
@@ -58,29 +61,39 @@ parent(a::AbstractArray) = a
 parentindexes(a::AbstractArray) = ntuple(i->OneTo(size(a,i)), ndims(a))
 
 ## SubArray creation
-# Drops singleton dimensions (those indexed with a scalar)
-function view{T,N}(A::AbstractArray{T,N}, I::Vararg{ViewIndex,N})
+# We always assume that the dimensionality of the parent matches the number of
+# indices that end up getting passed to it, so we store the parent as a
+# ReshapedArray view if necessary. The trouble is that `CartesianIndex`s make
+# the computation of the number of effective indices non-trivial. At this point
+# in the bootstrap, we don't have the required tools to do the computation,
+# so we just define `view` for non-CartesianIndex types here. The hard part is
+# done in multidimensional.jl after CartesianIndex gets defined properly.
+_maybe_reshape_parent(A::AbstractArray, ::NTuple{1, Bool}) = reshape(A, Val{1})
+_maybe_reshape_parent{_,N}(A::AbstractArray{_,N}, ::NTuple{N, Bool}) = A
+_maybe_reshape_parent{N}(A::AbstractArray, ::NTuple{N, Bool}) = reshape(A, Val{N}) # TODO: DEPRECATE FOR #14770
+function view(A::AbstractArray, I::ViewIndex...)
     @_inline_meta
     @boundscheck checkbounds(A, I...)
-    unsafe_view(A, I...)
-end
-function view(A::AbstractArray, i::ViewIndex)
-    @_inline_meta
-    @boundscheck checkbounds(A, i)
-    unsafe_view(reshape(A, Val{1}), i)
-end
-function view{N}(A::AbstractArray, I::Vararg{ViewIndex,N}) # TODO: DEPRECATE FOR #14770
-    @_inline_meta
-    @boundscheck checkbounds(A, I...)
-    unsafe_view(reshape(A, Val{N}), I...)
+    unsafe_view(_maybe_reshape_parent(A, index_ndims(I...)), I...)
 end
 
-function unsafe_view{T,N}(A::AbstractArray{T,N}, I::Vararg{ViewIndex,N})
+function unsafe_view(A::AbstractArray, I::ViewIndex...)
     @_inline_meta
     J = to_indexes(I...)
     SubArray(A, J, map(unsafe_length, index_shape(A, J...)))
 end
-function unsafe_view{T,N}(V::SubArray{T,N}, I::Vararg{ViewIndex,N})
+# Reindexing by CartesianIndex is hard, so flatten them out first
+unsafe_view(V::SubArray, I::ViewIndex...) = unsafe_view(V, IteratorsMD.flatten(I)...)
+unsafe_view(V::SubArray, I::Union{Real, NonSliceIndex}...) = (@_inline_meta; _maybe_reindex(V, to_indexes(I...)))
+# Reindexing by an array of cartesian indices is disastrously hard. So we punt.
+_maybe_reindex(V, I) = (@_inline_meta; _maybe_reindex(V, I, I))
+_maybe_reindex{C<:AbstractCartesianIndex}(V, I, ::Tuple{AbstractArray{C}, Vararg{Any}}) =
+    (@_inline_meta; SubArray(V, I, map(unsafe_length, index_shape(V, I...))))
```

Comment:
No, `view` will have already done the reshaping if any was necessary, and no conversion of indices is done here.  What's happening is that this wraps the parent view in a *second* SubArray.  Here's that example from above:

```jl
julia> A = copy(reshape(1:16, 4, 4))
       V = view(A, [1 2; 3 4], [4 2; 3 1])
       I = (:, map(CartesianIndex, [(1,1), (2, 2)]), [CartesianIndex()], :)
       V2 = view(V, I...)

julia> typeof(V2) # It's a subarray of a subarray; we weren't able to recompute the indices and "pop" the parent SubArray
SubArray{Int64,4,SubArray{Int64,4,Array{Int64,2},Tuple{Array{Int64,2},Array{Int64,2}},false},Tuple{Colon,Array{CartesianIndex{2},1},Array{CartesianIndex{0},1},Colon},false}

julia> V2.parent === V && V2.indexes === I
true
```